src/wasm/asm-wasm-builder.cc

```diff
@@ -0,0 +1,459 @@
+  RECURSE(Visit(expr->else_expression()));
+}
+
+
+void AsmWasmBuilder::VisitVariableProxy(VariableProxy* expr) {
+  if (in_function_) {
+    Variable* var = expr->var();
+    if (var->is_function()) {
+      std::vector<uint8_t> index =
+          UnsignedLEB128From(LookupOrInsertFunction(var));
+      current_function_builder_->AddBody(
+          index.data(), static_cast<uint32_t>(index.size()));
+    } else {
+      if (!is_set_op_) {
+        current_function_builder_->AddBody(kExprGetLocal, false);
+      }
+      std::vector<uint8_t> index =
+          UnsignedLEB128From(LookupOrInsertLocal(var, false));
+      uint32_t pos_of_index[1] = {0};
+      current_function_builder_->AddBody(
+          index.data(), static_cast<uint32_t>(index.size()), pos_of_index, 1);
+    }
+  } else if (marking_exported) {
+    Variable* var = expr->var();
+    if (var->is_function()) {
+      uint16_t index = LookupOrInsertFunction(var);
+      builder_->FunctionAt(index)->Exported(1);
+    }
+  }
+}
+
+
+void AsmWasmBuilder::VisitLiteral(Literal* expr) {
+  DCHECK(in_function_);
+  if (expr->raw_value()->IsNumber()) {
+    // TODO: deal with type
+    int val = static_cast<int>(expr->raw_value()->AsNumber());
+    current_function_builder_->AddBody(kExprI32Const, false);
+    current_function_builder_->AddBody(static_cast<byte>(val), false);
+    current_function_builder_->AddBody(static_cast<byte>(val >> 8), false);
+    current_function_builder_->AddBody(static_cast<byte>(val >> 16), false);
+    current_function_builder_->AddBody(static_cast<byte>(val >> 24), false);
+  }
+}
+
+
+void AsmWasmBuilder::VisitRegExpLiteral(RegExpLiteral* expr) {}
+
+
+void AsmWasmBuilder::VisitObjectLiteral(ObjectLiteral* expr) {
+  ZoneList<ObjectLiteralProperty*>* props = expr->properties();
+  for (int i = 0; i < props->length(); ++i) {
+    ObjectLiteralProperty* prop = props->at(i);
+    RECURSE(Visit(prop->value()));
+  }
+}
+
+
+void AsmWasmBuilder::VisitArrayLiteral(ArrayLiteral* expr) {
+  ZoneList<Expression*>* values = expr->values();
+  for (int i = 0; i < values->length(); ++i) {
+    Expression* value = values->at(i);
+    RECURSE(Visit(value));
+  }
+}
+
+void AsmWasmBuilder::VisitAssignment(Assignment* expr) {
+  DCHECK(in_function_);
+  current_function_builder_->AddBody(kExprSetLocal, false);
+  is_set_op_ = true;
+  RECURSE(Visit(expr->target()));
+  is_set_op_ = false;
+  RECURSE(Visit(expr->value()));
+}
+
+
+void AsmWasmBuilder::VisitYield(Yield* expr) {
+  RECURSE(Visit(expr->generator_object()));
+  RECURSE(Visit(expr->expression()));
+}
+
+
+void AsmWasmBuilder::VisitThrow(Throw* expr) {
+  RECURSE(Visit(expr->exception()));
+}
+
+
+void AsmWasmBuilder::VisitProperty(Property* expr) {
+  RECURSE(Visit(expr->obj()));
+  RECURSE(Visit(expr->key()));
+}
+
+
+void AsmWasmBuilder::VisitCall(Call* expr) {
+  DCHECK(in_function_);
+  current_function_builder_->AddBody(kExprCallFunction, false);
+  RECURSE(Visit(expr->expression()));
```

Comment:
You'll have to distinguish between direct calls, indirect calls, and calls to imported functions here.